scanpy/tools/_rank_genes_groups.py

```diff
@@ -18,18 +18,347 @@
 _CorrMethod = Literal['benjamini-hochberg', 'bonferroni']
 
 
+def _select_top_n(scores, n_top):
+    n_from = scores.shape[0]
+    reference_indices = np.arange(n_from, dtype=int)
+    partition = np.argpartition(scores, -n_top)[-n_top:]
+    partial_indices = np.argsort(scores[partition])[::-1]
+    global_indices = reference_indices[partition][partial_indices]
+
+    return global_indices
+
+
+def _ranks(X, mask=None, mask_rest=None):
+    CONST_MAX_SIZE = 10000000
+
+    n_genes = X.shape[1]
+
+    if issparse(X):
+        merge = lambda tpl: vstack(tpl).toarray()
+        adapt = lambda X: X.toarray()
+    else:
+        merge = np.vstack
+        adapt = lambda X: X
+
+    masked = mask is not None and mask_rest is not None
+
+    if masked:
+        n_cells = np.count_nonzero(mask) + np.count_nonzero(mask_rest)
+        get_chunk = lambda X, left, right: merge(
+            (X[mask, left:right], X[mask_rest, left:right])
+        )
+    else:
+        n_cells = X.shape[0]
+        get_chunk = lambda X, left, right: adapt(X[:, left:right])
+
+    # Now calculate gene expression ranking in chunkes:
+    chunk = []
+    # Calculate chunk frames
+    n_genes_max_chunk = floor(CONST_MAX_SIZE / n_cells)
+    if n_genes_max_chunk < n_genes:
+        chunk_index = n_genes_max_chunk
+        while chunk_index < n_genes:
+            chunk.append(chunk_index)
+            chunk_index = chunk_index + n_genes_max_chunk
+        chunk.append(n_genes)
+    else:
+        chunk.append(n_genes)
+
+    left = 0
+    # Calculate rank sums for each chunk for the current mask
+    for chunk_index, right in enumerate(chunk):
+
+        df = pd.DataFrame(data=get_chunk(X, left, right))
+        ranks = df.rank()
+        yield ranks, left, right
+
+        left = right
+
+
+class _RankGenes:
+    def __init__(self, adata, groups, groupby, reference, use_raw, layer):
```

Comment:
Could these values get documented? At the very least, what their types should be?